File: cbls/violation.py
"""Penalty objective F = f + λ·V with adaptive λ."""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cbls.model import Model
    from cbls.dag import ExprNode


class ViolationManager:
    """Manages constraint violations and augmented objective."""

    def __init__(self, model: Model):
        self.model = model
        self.weights: list[float] = [1.0] * len(model.constraints)
        self.adaptive_lambda = AdaptiveLambda()

    @property
    def lambda_(self) -> float:
        return self.adaptive_lambda.lambda_

    def constraint_violation(self, i: int) -> float:
        """Violation of constraint i: max(0, expr_i.value)."""
        return max(0.0, self.model.constraints[i].value)

    def total_violation(self) -> float:
        """V(x) = Σ max(0, constraint_i.value) * weight_i."""
        return sum(
            max(0.0, c.value) * w
            for c, w in zip(self.model.constraints, self.weights)
        )

    def augmented_objective(self) -> float:
        """F(x) = f(x) + λ · V(x)."""
        obj = self.model.objective.value if self.model.objective is not None else 0.0
        return obj + self.lambda_ * self.total_violation()

    def is_feasible(self, tol: float = 1e-9) -> bool:
        return all(c.value <= tol for c in self.model.constraints)

    def violated_constraints(self, tol: float = 1e-9) -> list[int]:
        return [i for i, c in enumerate(self.model.constraints) if c.value > tol]

    def bump_weights(self, factor: float = 1.0):
        """Increase weights on violated constraints (FJ-style)."""
        for i in self.violated_constraints():
            self.weights[i] += factor


class AdaptiveLambda:
    """Adapt λ based on search progress."""

    def __init__(self, initial_lambda: float = 1.0):
        self.lambda_: float = initial_lambda
        self.consecutive_infeasible: int = 0
        self.consecutive_feasible_stuck: int = 0

    def update(self, is_feasible: bool, obj_improved: bool):
        if not is_feasible:
            self.consecutive_infeasible += 1
            self.consecutive_feasible_stuck = 0
            if self.consecutive_infeasible > 10:
                self.lambda_ *= 1.5
                self.consecutive_infeasible = 0
        else:
            self.consecutive_infeasible = 0
            if not obj_improved:
                self.consecutive_feasible_stuck += 1
                if self.consecutive_feasible_stuck > 20:
                    self.lambda_ *= 0.8
                    self.consecutive_feasible_stuck = 0
            else:
                self.consecutive_feasible_stuck = 0

File: src/violation.cpp
#include "cbls/violation.h"

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace cbls {

// Non-convex objectives/constraints (exp/pow/div blowups — the MINLPLib target)
// can drive a node value to +inf or NaN. Such a value would poison the
// total_violation cache, the structural pass's move comparison and the
// best-objective bookkeeping. Map every non-finite (or absurdly large) violation
// to a large but finite penalty so the search treats the point as very bad but
// still well-ordered. The clamp is one-sided on the violation (which is already
// max(0, .)), so a finite-but-huge value and a +inf value both become kInfPenalty.
namespace {

double clamped_node_violation(double node_value) {
    // NaN must be handled before max(): std::max(0.0, NaN) returns 0.0, which
    // would silently mask a NaN constraint as satisfied. NaN (e.g. inf-inf,
    // 0*inf) is treated as a maximal violation — we have no evidence it holds.
    if (std::isnan(node_value)) {
        return kInfPenalty;
    }
    double v = std::max(0.0, node_value);
    if (v > kInfPenalty) {  // also catches +inf
        return kInfPenalty;
    }
    return v;
}
}  // namespace

ViolationManager::ViolationManager(Model& model) : model_(model) {
    weights.resize(model.constraint_ids().size(), 1.0);
    cached_violations_.resize(model.constraint_ids().size(), 0.0);
}

double ViolationManager::constraint_violation(int i) const {
    if (i < 0 || i >= static_cast<int>(model_.constraint_ids().size())) {
        throw std::out_of_range("constraint index out of range");
    }
    int32_t cid = model_.constraint_ids()[i];
    return clamped_node_violation(model_.node(cid).value);
}

void ViolationManager::recompute_cache() const {
    const auto& cids = model_.constraint_ids();
    cached_total_ = 0.0;
    for (size_t i = 0; i < cids.size(); ++i) {
        cached_violations_[i] = clamped_node_violation(model_.node(cids[i]).value);
        cached_total_ += cached_violations_[i] * weights[i];
    }
    cache_valid_ = true;
    incremental_updates_ = 0;
}

double ViolationManager::total_violation() const {
    if (!cache_valid_) {
        recompute_cache();
        return cached_total_;
    }

    // Periodically recompute from scratch to prevent floating-point drift
    if (++incremental_updates_ >= 1000) {
        recompute_cache();
        return cached_total_;
    }

    // Incremental update: check which constraints changed
    const auto& cids = model_.constraint_ids();
    for (size_t i = 0; i < cids.size(); ++i) {
        double new_viol = clamped_node_violation(model_.node(cids[i]).value);
        if (new_viol != cached_violations_[i]) {
            cached_total_ += (new_viol - cached_violations_[i]) * weights[i];
            cached_violations_[i] = new_viol;
        }
    }
    return cached_total_;
}

double ViolationManager::augmented_objective() const {
    double obj = 0.0;
    if (model_.objective_id() >= 0) {
        obj = model_.node(model_.objective_id()).value;
        if (!std::isfinite(obj)) {
            obj = kInfPenalty;  // non-convex blowup: keep the metric ordered
        }
    }
    return obj + total_violation();
}

void ViolationManager::snapshot_violations(std::vector<double>& out) const {
    // total_violation() brings cached_violations_ up to the current node values
    // (it diffs every constraint against the cache on each call), so the snapshot
    // is a copy of work that has to happen anyway rather than a second scan.
    total_violation();
    out = cached_violations_;
}

double ViolationManager::weighted_delta_from(const std::vector<double>& snapshot) const {
    const auto& cids = model_.constraint_ids();
    if (snapshot.size() != cids.size()) {
        throw std::invalid_argument("weighted_delta_from: snapshot size != constraint count");
    }
    // Per-constraint differencing, for the reason spelled out on
    // Model::weighted_violation_delta: `1e30 + 3` rounds back to `1e30`, so
    // subtracting two whole sums loses every real row the moment one row is
    // clamped, whereas `1e30 - 1e30` cancels to exactly 0 here.
    //
    // Deliberately does not update the cache: the caller's candidate move is
    // usually rolled back, and total_violation() self-corrects against whatever
    // the node values are when it is next called either way.
    double delta = 0.0;
    for (size_t i = 0; i < cids.size(); ++i) {
        const double now = clamped_node_violation(model_.node(cids[i]).value);
        if (now != snapshot[i]) {
            delta += weights[i] * (now - snapshot[i]);
        }
    }
    return delta;
}

bool ViolationManager::is_feasible(double tol) const {
    for (int32_t cid : model_.constraint_ids()) {
        // A NaN node value is not <= tol; treat it as infeasible (the `!(<=)`
        // form catches NaN, which a bare `> tol` would silently pass).
        if (!(model_.node(cid).value <= tol)) {
            return false;
        }
    }
    return true;
}

std::vector<int> ViolationManager::violated_constraints(double tol) const {
    std::vector<int> result;
    const auto& cids = model_.constraint_ids();
    for (size_t i = 0; i < cids.size(); ++i) {
        if (!(model_.node(cids[i]).value <= tol)) {
            result.push_back(static_cast<int>(i));
        }
    }
    return result;
}

void ViolationManager::bump_weights(double factor) {
    for (int i : violated_constraints()) {
        weights[i] += factor;
    }
    cache_valid_ = false;  // weights changed, invalidate
}

double ViolationManager::weighted_violation_delta(int32_t var_id, double j) const {
    // Delegate to the allocation-free Model probe (hot path: one call per jump
    // candidate). weights is the per-constraint GLS weight vector.
    return model_.weighted_violation_delta(var_id, j, weights);
}

}  // namespace cbls
